File: sort/quicksort.js
function part(array, low, max){
   var pivot = array[max]
   var i = low-1
   for(var j = low; j <= max-1;j++){
       if(array[j] < pivot){
           i ++
           [array[i], array[j]] = [array[j], array[i]]
           }
    }
    [array[i+1], array[max]] = [array[max], array[i+1]]
    return i+1
}
function quicksort(arr, low, high){
    if(low < high){
        p = part(arr, low, high)
        quicksort(arr,low, p-1);
        quicksort(arr, p+1, high);
    }
}
var arr = [5,6,3,7,0,0,6,5,4,7]
quicksort(hm, 0, arr.length-1)

File: sort/quicksort.cpp
#include <cstdlib>
#include <algorithm>

template<typename TP, int N>
int part(TP (&arr)[N], int low, int max){
    int pivot = arr[max];
    int i = low-1;
    for(int j = low; j <= max; ++j){
        if(arr[j] < pivot){
            ++i;
            std::swap(arr[i], arr[j]);
        }
    }
    std::swap(arr[i+1], arr[max]);
    return i+1;
}

template<typename TP, int N>
void quick_sort(TP (&arr)[N], int low, int max){
    if(low < max){
        int p = part(arr, low, max);
        quick_sort(arr, low, p-1);
        quick_sort(arr, p+1, max);
    }
}
